sklearn/utils/extmath.py

```diff
@@ -743,12 +749,21 @@ def cartesian(arrays, out=None):
     """
     arrays = [np.asarray(x) for x in arrays]
     shape = (len(x) for x in arrays)
-    dtype = arrays[0].dtype
 
     ix = np.indices(shape)
     ix = ix.reshape(len(arrays), -1).T
 
     if out is None:
+        # find the most permissive dtype among all input arrays
+        dtypes = list({x.dtype for x in arrays})
+        can_cast = [
+            np.can_cast(from_, to_, casting="safe")
+            for from_, to_ in itertools.product(dtypes, repeat=2)
+        ]
+        # The most permissive dtype is the one that has a column full of True
+        can_cast = np.reshape(can_cast, (len(dtypes), len(dtypes))).all(axis=0)
+        dtype = dtypes[can_cast.argmax()]
```

Comment:
Maybe we can reuse [`np.common_type`](https://numpy.org/doc/stable/reference/generated/numpy.common_type.html#numpy.common_type)? or [`np.find_common_type`](https://numpy.org/doc/stable/reference/generated/numpy.find_common_type.html)

Reply:
Since `find_common_type` is going to be deprecated in NumPy: https://github.com/numpy/numpy/pull/22539, I prefer `np.result_type`. (`common_type` does not work with object dtypes).

Reply:
`np.result_type` seems like the right fit.